### ISSUE TITLE
BUG: AVX-512 results differ from non-AVX-512 for sin, cos, exp, log, etc.

### ISSUE THREAD

**Issue by asreimer** [numpy/numpy#23523]
### Describe the issue:

It appears that there is a small difference between functions executed using AVX-512 and not using AVX-512. Is this expected? My admittedly limited understanding is that computation with AVX-512 should produce results identical to other computations.

For example, `math.sin` and `np.sin` agree with AVX-512 disabled, but disagree by approximately`2**-53` depending on the input angle. `np.cos` is similar. I've also tested some of the other SIMD backed numpy functions, such as `np.exp` and `np.log` and they also return non-zero differences with the `math` functions unless AVX-512 is explicitly disabled or not available (such as on AMD or ARM).

### Reproduce the code example:

```python
(py39) [asreimer@fedora ~]$ python
Python 3.9.16 (main, Apr  3 2023, 12:24:01) 
[GCC 11.3.1 20220421 (Red Hat 11.3.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import math
>>> import numpy as np
>>> print('Difference of sines: {}'.format(np.sin(np.radians(45))-math.sin(np.radians(45))))
Difference of sines: 1.1102230246251565e-16
>>> print('Difference of exponentials: {}'.format(np.exp(12.312)-math.exp(12.312)))
Difference of exponentials: -2.9103830456733704e-11
>>> print('Difference of natural logorithms: {}'.format(np.log(1332.312)-math.log(1332.312)))
Difference of natural logorithms: -8.881784197001252e-16
>>> 
(py39) [asreimer@fedora ~]$ NPY_DISABLE_CPU_FEATURES="AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL" python
Python 3.9.16 (main, Apr  3 2023, 12:24:01) 
[GCC 11.3.1 20220421 (Red Hat 11.3.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import math
>>> import numpy as np
>>> print('Difference of sines: {}'.format(np.sin(np.radians(45))-math.sin(np.radians(45))))
Difference of sines: 0.0
>>> print('Difference of exponentials: {}'.format(np.exp(12.312)-math.exp(12.312)))
Difference of exponentials: 0.0
>>> print('Difference of natural logorithms: {}'.format(np.log(1332.312)-math.log(1332.312)))
Difference of natural logorithms: 0.0
>>>
```

### Runtime information:

Default:

    >>> import sys, numpy; print(numpy.__version__); print(sys.version); print(numpy.show_runtime())
    1.24.2
    3.9.16 (main, Apr  3 2023, 12:24:01) 
    [GCC 11.3.1 20220421 (Red Hat 11.3.1-3)]
    [{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                          'found': ['SSSE3',
                                    'SSE41',
                                    'POPCNT',
                                    'SSE42',
                                    'AVX',
                                    'F16C',
                                    'FMA3',
                                    'AVX2',
                                    'AVX512F',
                                    'AVX512CD',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL'],
                          'not_found': ['AVX512_KNL', 'AVX512_KNM']}},
     {'architecture': 'SkylakeX',
      'filepath': '/home/asreimer/venvs/py39/lib/python3.9/site-packages/numpy.libs/libopenblas64_p-r0-15028c96.3.21.so',
      'internal_api': 'openblas',
      'num_threads': 4,
      'prefix': 'libopenblas',
      'threading_layer': 'pthreads',
      'user_api': 'blas',
      'version': '0.3.21'}]

With AVX-512 disabled via `NPY_DISABLE_CPU_FEATURES`:

    >>> import sys, numpy; print(numpy.__version__); print(sys.version); print(numpy.show_runtime())
    1.24.2
    3.9.16 (main, Apr  3 2023, 12:24:01) 
    [GCC 11.3.1 20220421 (Red Hat 11.3.1-3)]
    [{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                          'found': ['SSSE3',
                                    'SSE41',
                                    'POPCNT',
                                    'SSE42',
                                    'AVX',
                                    'F16C',
                                    'FMA3',
                                    'AVX2'],
                          'not_found': ['AVX512F',
                                        'AVX512CD',
                                        'AVX512_KNL',
                                        'AVX512_KNM',
                                        'AVX512_SKX',
                                        'AVX512_CLX',
                                        'AVX512_CNL',
                                        'AVX512_ICL']}},
     {'architecture': 'SkylakeX',
      'filepath': '/home/asreimer/venvs/py39/lib/python3.9/site-packages/numpy.libs/libopenblas64_p-r0-15028c96.3.21.so',
      'internal_api': 'openblas',
      'num_threads': 4,
      'prefix': 'libopenblas',
      'threading_layer': 'pthreads',
      'user_api': 'blas',
      'version': '0.3.21'}]

### Context for the issue:

While the differences here are small, they unexpected and can cause larger problems. For example, when performing vector rotations using rotation matrices, the small differences of 1e-16 can blow up to something closer to 1e-5. This is how I originally found this bug: finding differences in expected output while using rotation matrices on AMD and ARM based systems without AVX-512 compared to the results from an Intel system with AVX-512.

**Comment by asreimer** [comment]
Note that this issue is not present in `numpy<=1.20.3`.

**Comment by mattip** [comment]
This is a speed-vs-consistency tradeoff. We accept differences of up to 4 ULP from the "golden results", depending on which implementations are used. You can use [`NPY_DISABLE_CPU_FEATURES`](https://numpy.org/devdocs/reference/simd/build-options.html#runtime-dispatch) and, once #22058 is merged, `NPY_ENABLE_CPU_FEATURES` to get consistent answers across different machines.

> Note that this issue is not present in numpy<=1.20.3.

I think older versions do not have SIMD accelerated loops.

**Comment by asreimer** [comment]
Thank you, I appreciate the reply! For now, as you can see in the logs I posted, we have already been using the `NPY_DISABLE_CPU_FEATURES` variable. That was how I was able to confirm this issue was specific to AVX-512.

I really like the idea of #22058. Making SIMD opt-out when it breaks consistency causes headaches, especially when the release notes don't include warnings or notes about this being the case (maybe they do? I couldn't find anything searching yesterday, but I could have overlooked something, it was a long day). I say this not to be critical, but to add some weight to making this an opt-in feature. Users should knowingly opt-in to making speed-vs-consistency tradeoffs like this, but I know this is a tricky line to walk as a developer.

**Comment by asreimer** [comment]
I would be happy to close this when https://github.com/numpy/numpy/pull/22137 is merged. I think that is a sensible solution to this issue.

**Comment by zmbc** [comment]
I think this can be closed now, but it is still not documented, see https://github.com/numpy/numpy/issues/22059

**Comment by zmbc** [comment]
I agree with @asreimer that I wish this was opt-in. I could even imagine it printing a warning when these features are available but _not_ opted into, but the current opt-out default is surprising to me.

I bet it is surprising to a lot of other people too, but the reason they aren't commenting here is because they never figure out the root cause of why they keep getting different results every time they run their code on a different computer. It took me quite some time to debug where in a complex data processing pipeline the discrepancies were sneaking in, and I did _not_ expect it to be `np.log`. `np.log` is a trivial amount of the computation I'm doing and turning off the optimizations doesn't cause even a perceptible performance difference for my overall task.

**Comment by seberg** [comment]
> I think this can be closed now, but it is still not documented

@zmbc actually, we removed the biggest precision offenders again (not sure if in NumPy 1.25 or 1.26), but not sure how exact a match you are hoping for.

FWIW, even *without* any SIMD code there are many things that are expected to differ from platform to platform (hardware *or* software difference, like OS/math library versions):  it's a deep stack and computers don't make it easy to get the exact same results.

**Comment by zmbc** [comment]
@seberg I saw this on 1.26.4.

> not sure how exact a match you are hoping for.

*Exact.* Same bits.

> FWIW, even without any SIMD code there are many things that are expected to differ from platform to platform (hardware or software difference, like OS/math library versions): it's a deep stack and computers don't make it easy to get the exact same results.

Sure, it's not _easy_. But it is possible, and very very useful in computational pipelines for reproducible science. I have been able to get very large, very complex pipelines to be completely bitwise reproducible across a wide range of CPUs, using features like MKL's [CNR](https://www.intel.com/content/www/us/en/developer/articles/technical/introduction-to-the-conditional-numerical-reproducibility-cnr.html). I think it is very reasonable to request that NumPy have similar features. If you decide that this isn't a priority, that is fine -- but it absolutely is feasible and could be done if you wanted to.
